fw/Core/Hitcon/Logic/XBoardLogic.cc

```diff
@@ -0,0 +1,185 @@
+#include <Logic/XBoardLogic.h>
+
+#include <cstring>
+
+using namespace hitcon::service::sched;
+using namespace hitcon::service::xboard;
+
+namespace hitcon {
+namespace service {
+namespace xboard {
+
+XBoardLogic g_xboard_logic;
+
+constexpr size_t PKT_PAYLOAD_LEN_MAX = 256;
+static uint8_t packet_payload[PKT_PAYLOAD_LEN_MAX];
+
+struct Frame {
+  uint64_t preamble;  // 0xD555555555555555
+  uint16_t id;
+  uint8_t len;
+  uint8_t type;  // 8: ping
+  uint16_t checksum;
+};
+constexpr size_t HEADER_SZ = sizeof(Frame);
+
+uint8_t tx_buf[] = "Hello World";
+uint8_t rx_buf[RX_BUF_SZ] = {0};
+static uint16_t prod_head = 0;
+static uint16_t cons_head = 0;
+static bool recv_ping = false;
+static uint8_t no_ping_count = 0;
+
+uint8_t alive_message[] = "alive_message";
+void send_ping() {
+  uint8_t pkt[sizeof(Frame)] = {0};
+  *(Frame *)pkt = Frame{0xD555555555555555, 0, 0, 8, 0};
+  // for (int i = 0; i < sizeof(Frame); i++) {
+  //   pkt[i] = (0x11+i)&0x0FF;
+  //   pkt[i] = 0;
+  // }
+  g_xboard_service.QueueDataForTx(pkt, sizeof(pkt));
+}
+
+inline uint16_t inc_head(size_t head, size_t offset) {
+  return (head + offset) % RX_BUF_SZ;
+}
+
+// read `size` bytes from `rx_buf` to `dst`
+// if `head_offset` > 0, start reading from cons_head + head_offset
+// return false if no enough bytes to read
+bool try_read_bytes(uint8_t *dst, size_t size, uint16_t head_offset = 0) {
+  uint16_t _cons_head = inc_head(cons_head, head_offset);
+  if (cons_head == prod_head) {
+    if (_cons_head != cons_head) return false;
+  } else if (cons_head < prod_head) {
+    // Ok: c <= _c <= p
+    if (_cons_head > prod_head) return false;
+    if (_cons_head < cons_head) return false;
+  } else {
+    // Ok: p < c <= _c
+    // Ok: _c <= p < c
+    if (prod_head < _cons_head && _cons_head < cons_head) return false;
+  }
+  uint16_t remain_size =
+      (_cons_head > prod_head ? 0 : RX_BUF_SZ) - prod_head + _cons_head;
+  if (remain_size < size) {
+    return false;
+  }
+  uint16_t next_cons_head = inc_head(_cons_head, size);
+  if (next_cons_head < _cons_head) {
+    uint16_t sz1 = RX_BUF_SZ - _cons_head;
+    memcpy(dst, rx_buf + _cons_head, sz1);
+    memcpy(dst + sz1, rx_buf, size - sz1);
+  } else {
+    memcpy(dst, rx_buf + _cons_head, size);
+  }
+  return true;
+}
+
+void XBoardLogic::ParsePacket() {
+  recv_ping = false;
+  while (cons_head != prod_head) {
```

Comment:
I would recommend:
- Consider limiting the number of bytes processed or number of Frame processed in a single call because we don't want a task to run for too long.
- Add a variable that keeps track of up to which byte have we checked, so we don't need to check them again the next time.